src/vmm/src/arch/aarch64/pvtime.rs

```diff
@@ -0,0 +1,209 @@
+        ))
+    }
+
+    /// Create a new PVTime device given a base addr
+    /// - Assumes total shared memory region from base addr is already allocated
+    fn from_base(base_ipa: GuestAddress, vcpu_count: u8) -> Self {
+        PVTime {
+            vcpu_count,
+            base_ipa,
+        }
+    }
+
+    /// Creates a new PVTime device by allocating new system memory for all vCPUs
+    pub fn new(
+        resource_allocator: &mut ResourceAllocator,
+        vcpu_count: u8,
+    ) -> Result<Self, PVTimeError> {
+        // This returns the IPA of the start of our shared memory region for all vCPUs.
+        let base_ipa: GuestAddress = GuestAddress(
+            resource_allocator
+                .allocate_system_memory(
+                    STEALTIME_STRUCT_MEM_SIZE * vcpu_count as u64,
+                    STEALTIME_STRUCT_MEM_SIZE,
+                    vm_allocator::AllocPolicy::LastMatch,
+                )
+                .map_err(PVTimeError::AllocationFailed)?,
+        );
+        Ok(Self::from_base(base_ipa, vcpu_count))
+    }
+
+    /// Check if PVTime is supported on vcpu
+    pub fn is_supported(vcpu_fd: &VcpuFd) -> bool {
+        // Check if pvtime is enabled
+        let pvtime_device_attr = kvm_bindings::kvm_device_attr {
+            group: kvm_bindings::KVM_ARM_VCPU_PVTIME_CTRL,
+            attr: kvm_bindings::KVM_ARM_VCPU_PVTIME_IPA as u64,
+            addr: 0,
+            flags: 0,
+        };
+
+        // Use kvm_has_device_attr to check if PVTime is supported
+        vcpu_fd.has_device_attr(&pvtime_device_attr).is_ok()
+    }
+
+    /// Register a vCPU with its pre-allocated steal time region
+    fn register_vcpu(
+        &self,
+        vcpu_index: u8,
+        vcpu_fd: &kvm_ioctls::VcpuFd,
+    ) -> Result<(), PVTimeError> {
+        // Get IPA of the steal_time region for this vCPU
+        let ipa = self.get_steal_time_region_addr(vcpu_index)?;
+
+        // Use KVM syscall (kvm_set_device_attr) to register the vCPU with the steal_time region
+        let vcpu_device_attr = kvm_bindings::kvm_device_attr {
+            group: KVM_ARM_VCPU_PVTIME_CTRL,
+            attr: KVM_ARM_VCPU_PVTIME_IPA as u64,
+            addr: &ipa.0 as *const u64 as u64, // userspace address of attr data
+            flags: 0,
+        };
+
+        vcpu_fd
+            .set_device_attr(&vcpu_device_attr)
+            .map_err(PVTimeError::DeviceAttribute)?;
+
+        Ok(())
+    }
+
+    /// Register all vCPUs with their pre-allocated steal time regions
+    pub fn register_all_vcpus(&self, vcpus: &mut [Vcpu]) -> Result<(), PVTimeError> {
+        // Register the vcpu with the pvtime device to map its steal time region
+        for (i, vcpu) in vcpus.iter().enumerate() {
+            #[allow(clippy::cast_possible_truncation)]
+            // We know vcpu_count is u8 according to VcpuConfig
+            self.register_vcpu(i as u8, &vcpu.kvm_vcpu.fd)?;
```

Comment:
nit: I think a `try_into` would be nicer
```suggestion
            self.register_vcpu(i.try_into().expect("cpu index should be less than 256"), &vcpu.kvm_vcpu.fd)?;
```

Reply:
Avoiding this by using a u64 for vcpu_index